pueblo/ngr/runner.py

```diff
@@ -455,22 +453,66 @@ def test(self) -> None:
         """
         Test a Python thing, based on which test runner is installed.
 
-        TODO: Figure out how to invoke target `poe check`, which is popular amongst
-              users of `pueblo`. It bundles linter/checkstyle and software test targets
-              into a single entrypoint, similar to what `gradle check` is doing.
+        First, try a high-level target of some sort of management tool,
+        discover its targets, and invoke `check` or `test`.
+
+        When no high-level tool can be discovered, just invoke `pytest`.
         """
 
+        # On Poetry projects, invoke poetry/pytest.
+        # TODO: Also allow using Poetry with other task runners, see below.
         if self.has_poetry_lock:
             # TODO: poetry run which pytest
             run_command("poetry run pytest --config-file=.")
 
         else:
+            uses_poe = self.has_pyproject_toml and self.pyproject_has_string("[tool.poe.tasks]")
             has_pytest = shutil.which("pytest") is not None
-            if has_pytest:
+
+            # When a `pyproject.toml` file includes `poethepoet` task definitions, try to
+            # invoke sensible targets like `check` or `test`. This is popular amongst users
+            # of `pueblo` and friends.
+            # `poe check` bundles linter/checkstyle and software test targets into a single
+            # entrypoint, similar to how `gradle check` is doing it.
+            if uses_poe:
+                candidates = ["check", "test"]
+                poe_tasks = self.get_poe_tasks()
+                success = False
+                for candidate in candidates:
+                    if candidate in poe_tasks:
+                        run_command(f"poe {candidate}")
+                        success = True
+                        break
+                if not success:
+                    raise RuntimeError(f"Failed to discover poe task from candidates: {candidates}")
+
+            elif has_pytest:
                 run_command("pytest")
             else:
                 raise NotImplementedError("No handler to invoke Python item")
 
+    def pyproject_has_string(self, needle: str) -> bool:
+        """
+        Check whether `pyproject.toml` file contains given string.
+        """
+        pyproject_toml_file = self.path / "pyproject.toml"
+        pyproject_toml = pyproject_toml_file.read_text()
+        return needle in pyproject_toml
+
+    def get_poe_tasks(self) -> t.List[str]:
+        """
+        From `poethepoet._list_tasks`.
+        """
+        try:
+            from poethepoet.config import PoeConfig
+
+            config = PoeConfig()
+            config.load()
+            return [task for task in config.tasks.keys() if task and task[0] != "_"]
+        except Exception as ex:  # pylint: disable=broad-except
+            # this happens if there's no pyproject.toml present
```

Comment:
```suggestion
            # this happens if there's no pyproject.toml or task definition present
```

Reply:
I've copied that code from `poethepot`. I think at this very spot, the comments holds true, that it will only raise an exception when no `pyproject.toml` file is present. However, I haven't validated it.

Reply:
I've just validated it: The code returns an empty list, when no tasks are defined, i.e. when only an empty `[tool.poe.tasks]` section is defined.